[OpenMP] Adds omp_target_is_accessible routine

nicebert opened the pull request:
Adds omp_target_is_accessible routine.
Refactors common code from omp_target_is_present to work for both routines.

nicebert committed:
[OpenMP] Adds omp_target_is_accessible routine

Adds implementation of omp_target_is_accessible routine with
5.1 behaviour, checking if a host pointer is acccessible from a device
without running on the device (from the host).

nicebert committed:
[OpenMP] Adds omp_target_is_accessible routine

Adds implementation of omp_target_is_accessible routine with
5.1 behaviour, checking if a host pointer is acccessible from a device
without running on the device (from the host).

nicebert committed:
[OpenMP] Adds omp_target_is_accessible routine

Adds implementation of omp_target_is_accessible routine with
5.1 behaviour, checking if a host pointer is acccessible from a device
without running on the device (from the host).

nicebert committed:
[OpenMP] Adds omp_target_is_accessible routine

Adds implementation of omp_target_is_accessible routine with
5.1 behaviour, checking if a host pointer is acccessible from a device
without running on the device (from the host).

nicebert committed:
[OpenMP] Adds omp_target_is_accessible routine

Adds implementation of omp_target_is_accessible routine with
5.1 behaviour, checking if a host pointer is acccessible from a device
without running on the device (from the host).

shiltian commented on offload/libomptarget/OpenMP/API.cpp:
```diff
@@ -168,33 +170,25 @@ EXTERN int omp_target_is_present(const void *Ptr, int DeviceNum) {
+     " with size %zu\n",
+     DeviceNum, DPxPTR(Ptr), Size);
+
+  // omp_target_is_accessible tests whether a host pointer refers to storage
```
Does this "accessible" mean direct access or even map counts?

nicebert replied:
Thanks for pointing it out. I didn't compare the wording to "is_present" carefully enough and made a wrongful assumption based on the wording for "accessible". I'll rework the implementation.

shiltian commented on offload/libomptarget/device.cpp:
```diff
@@ -281,3 +281,7 @@ bool DeviceTy::useAutoZeroCopy() {
+
+bool DeviceTy::supportsUnifiedMemory() {
+  return RTL->supports_unified_memory(RTLDeviceID);
+}
```
missing empty line at EoF

shiltian commented on offload/libomptarget/OpenMP/API.cpp:
```diff
@@ -195,6 +195,49 @@ EXTERN int omp_target_is_present(const void *Ptr, int DeviceNum) {
   return Rc;
 }
 
+/// Check whether a pointer is accessible from a device.
+/// the functionality is available in OpenMP 5.1 and later
+/// OpenMP 5.1
+/// omp_target_is_accessible checks whether a host pointer is accessible from a
+/// device OpenMP 6.0 removes restriction on pointer, allowing any pointer
+/// interpreted as a pointer in the address space of the given device.
```
Can you re-format this entire paragraph? It seems like the line breaks are pretty random.

nicebert replied:
I ran clang format over it and that was the result.

shiltian commented on offload/include/omptarget.h:
```diff
@@ -280,6 +280,7 @@ int omp_get_initial_device(void);
 void *omp_target_alloc(size_t Size, int DeviceNum);
 void omp_target_free(void *DevicePtr, int DeviceNum);
 int omp_target_is_present(const void *Ptr, int DeviceNum);
+int omp_target_is_accessible(const void *Ptr, size_t Size, int DeviceNum);
```
You will also need to add the entry in `libomp` if I'm not mistaken.

mjklemm replied:
Yep.  I think there need to be entries in `omp_lib.h.var`, `omp.h.var`, `omp_lib.F90.var`.  You may also need to add something in `kmp_ftn_os.h`  and `dllexports` (but that's only for stubs).

nicebert replied:
those entries were already present in the compiler

jprotze replied:
Entries in kmp_ftn_entry.h and kmp_ftn_os.h are missing. Check `FTN_TARGET_IS_PRESENT` for reference.

@mjklemm at the moment, I cannot successfully compile a program using any of the device memory information routines without `-fopenmp-targets` flag. Is this a bug or feature?

shiltian commented on offload/libomptarget/OpenMP/API.cpp:
```diff
@@ -195,6 +195,49 @@ EXTERN int omp_target_is_present(const void *Ptr, int DeviceNum) {
+    return true;
+  }
+
+  // the device number must refer to a valid device
```
```suggestion
  // The device number must refer to a valid device
```

shiltian commented on offload/libomptarget/OpenMP/API.cpp:
```diff
@@ -195,6 +195,49 @@ EXTERN int omp_target_is_present(const void *Ptr, int DeviceNum) {
+  if (!DeviceOrErr)
+    FATAL_MESSAGE(DeviceNum, "%s", toString(DeviceOrErr.takeError()).c_str());
+
+  // for OpenMP 5.1 the routine checks whether a host pointer is accessible from
```
```suggestion
  // For OpenMP 5.1 the routine checks whether a host pointer is accessible from
```

shiltian commented on offload/libomptarget/OpenMP/API.cpp:
```diff
@@ -195,6 +195,49 @@ EXTERN int omp_target_is_present(const void *Ptr, int DeviceNum) {
+    FATAL_MESSAGE(DeviceNum, "%s", toString(DeviceOrErr.takeError()).c_str());
+
+  // for OpenMP 5.1 the routine checks whether a host pointer is accessible from
+  // the device this requires for the device to support unified shared memory
```
ditto

shiltian commented on offload/libomptarget/OpenMP/API.cpp:
```diff
@@ -195,6 +195,49 @@ EXTERN int omp_target_is_present(const void *Ptr, int DeviceNum) {
+    return true;
+  }
+
+  // functionality to check whether a device pointer is accessible from a device
```
ditto

shiltian commented on offload/libomptarget/OpenMP/API.cpp:
```diff
@@ -195,6 +195,49 @@ EXTERN int omp_target_is_present(const void *Ptr, int DeviceNum) {
 
+/// Check whether a pointer is accessible from a device.
+/// the functionality is available in OpenMP 5.1 and later
+/// OpenMP 5.1
```
How do we know the OpenMP version used by the user?

mjklemm replied:
I do not know of any way that the library could tell what OpenMP version has been requested in the compiler.  We would have to implement an infrastructure for this first and then check in the routines.  Alternatively, we could version the symbol and if the symbol is from a newer OpenMP version that what the compiler requested, fail in the linker.

kparzysz replied:
The runtime shouldn't care about the version.  The compiler can check the version, and can warn/error out when it sees a call to `omp_xyz` routine that is unavailable in that version (IIRC identifiers starting with omp_ are reserved for the implementation).

mjklemm replied:
Well, this is more about changed/extended behavior of a routine like this one.  There is a material extension of the behavior with version 6.0 over the 5.2 semantics of this routine.

shiltian replied:
That's right, so I'm wondering how we can tell here.

mjklemm commented on offload/libomptarget/exports:
```diff
@@ -37,6 +37,7 @@ VERS1.0 {
     __kmpc_push_target_tripcount_mapper;
     ompx_dump_mapping_tables;
     omp_get_mapped_ptr;
+    omp_target_is_accessible;
```
Please move this down in the list (alphabetically sorted?)

nicebert committed:
Update offload/libomptarget/OpenMP/API.cpp

Co-authored-by: Shilei Tian <i@tianshilei.me>

nicebert committed:
Update offload/libomptarget/OpenMP/API.cpp

Co-authored-by: Shilei Tian <i@tianshilei.me>

nicebert committed:
Fix comment spelling

Co-authored-by: Shilei Tian <i@tianshilei.me>

nicebert committed:
Fix comment spelling

Co-authored-by: Shilei Tian <i@tianshilei.me>

nicebert committed:
Merge branch 'feat/omp_target_is_accessible' of github.com:nicebert/llvm-project into feat/omp_target_is_accessible

nicebert committed:
Merge branch 'main' into feat/omp_target_is_accessible

jprotze commented on offload/libomptarget/OpenMP/API.cpp:
```diff
@@ -93,7 +93,8 @@ EXTERN int omp_get_device_num(void) {
   TIMESCOPE();
   OMPT_IF_BUILT(ReturnAddressSetterRAII RA(__builtin_return_address(0)));
-  int HostDevice = omp_get_num_devices();
+  int NumDevices = omp_get_num_devices();
+  int HostDevice = NumDevices == 0 ? -1 : NumDevices;
```
The application will always see the value provided from libomp, which is 0 if no other device is present: https://github.com/llvm/llvm-project/blob/c05611f1917ca01371e7f7915f4c94055c5a19cf/openmp/runtime/src/kmp_ftn_entry.h#L1197

Only the get_num_devices call is deligated to libomptarget

nicebert committed:
[OpenMP] Reverts omp_get_initial_device changes

nicebert committed:
[OpenMP] Reverts omp_get_initial_device changes

nicebert committed:
Merge branch 'feat/omp_target_is_accessible' of github.com:nicebert/llvm-project into feat/omp_target_is_accessible

nicebert committed:
Merge branch 'main' into feat/omp_target_is_accessible

nicebert committed:
Merge branch 'main' into feat/omp_target_is_accessible

mjklemm commented on offload/libomptarget/OpenMP/API.cpp:
```diff
@@ -195,6 +196,48 @@ EXTERN int omp_target_is_present(const void *Ptr, int DeviceNum) {
+    return false;
+  }
+
+  if (DeviceNum == omp_get_initial_device()) {
```
Please do the correct check in this routine.  I agree with your position of creating a new PR for fixing up the other routines.  Can you please record that as an issue?

nicebert replied:
added the check against -1 for this routine

nicebert committed:
[OpenMP] Rework implementation to be conform to OpenMP 6.0

Removes restriction on Ptr to host pointers.
Provides accessibility check for AMDGPU through HSA + default implementation returning false.

nicebert committed:
Merge branch 'feat/omp_target_is_accessible' of github.com:nicebert/llvm-project into feat/omp_target_is_accessible

nicebert committed:
Applies git-clang-format

nicebert commented:
> PR is broken and touches >5000 files

I think you're right. I must've messed up somewhere while rebasing

nicebert committed:
Merge branch 'main' into feat/omp_target_is_accessible

nicebert commented:
@carlobertolli can you review this. I've reworked the patch using hsa_amd_pointer_info as we discussed.

@CatherineMoore FIY

nicebert committed:
Fixes formatting and comment issues.

nicebert committed:
Merge branch 'feat/omp_target_is_accessible' of github.com:nicebert/llvm-project into feat/omp_target_is_accessible

nicebert committed:
Merge branch 'main' into feat/omp_target_is_accessible

nicebert committed:
[OpenMP] Fixes check for host device number

The implemetation is allowed to return -1 for the host device number.
To be complient with the spec both the device number needs to be checked against both -1
as well as the value returned by omp_get_initial_device.

nicebert committed:
Merge branch 'feat/omp_target_is_accessible' of github.com:nicebert/llvm-project into feat/omp_target_is_accessible

nicebert committed:
Fixes formatting

nicebert committed:
Merge branch 'main' into feat/omp_target_is_accessible

nicebert committed:
Merge branch 'main' into feat/omp_target_is_accessible

shiltian commented on offload/plugins-nextgen/amdgpu/src/rtl.cpp:
```diff
@@ -3027,6 +3027,29 @@ struct AMDGPUDeviceTy : public GenericDeviceTy, AMDGenericDeviceTy {
+    hsa_status_t Status =
+        hsa_amd_pointer_info(Ptr, &Info, malloc, &Count, &Agents);
+
+    if (Status != HSA_STATUS_SUCCESS)
```
Can we log here what is the reason of failing?

nicebert replied:
Added some handling for this with the latest updates.

carlobertolli reviewed (approved):
With the requested changes, this patch is, in my opinion, ready to land.

nicebert committed:
Review nit add braces to for loop body

Co-authored-by: Shilei Tian <i@tianshilei.me>

nicebert committed:
[OpenMP] Changes error handling for omp_target_is_accessible

Addresses rewiev comments.
Changes LIT test to XFAIL on nvptx.
Changes implementation to provide better debug information to the user about failures.

nicebert committed:
Merge branch 'feat/omp_target_is_accessible' of github.com:nicebert/llvm-project into feat/omp_target_is_accessible

nicebert committed:
Merge branch 'main' into feat/omp_target_is_accessible

nicebert commented on offload/libomptarget/OpenMP/API.cpp:
```diff
@@ -196,6 +196,34 @@ EXTERN int omp_target_is_present(const void *Ptr, int DeviceNum) {
+  // The device number must refer to a valid device
+  auto DeviceOrErr = PM->getDevice(DeviceNum);
+  if (!DeviceOrErr)
+    FATAL_MESSAGE(DeviceNum, "%s", toString(DeviceOrErr.takeError()).c_str());
```
I think this is a discussion we need to have in another setting than this issue as it does not just affect this API routine but existing routines (of which I stated a couple of examples above). My implementation follows the way this is currently handled so I think if you want to change the handling the discussion needs to be had in a different setting with a bigger audience.

shiltian replied:
Fair enough. CC @jhuber6 Maybe a topic for the next Wed meeting. :-)

mjklemm reviewed (approved):
LGTM

nicebert committed:
Merge branch 'main' into feat/omp_target_is_accessible

nicebert committed:
Merge branch 'main' into feat/omp_target_is_accessible

nicebert committed:
Merge branch 'main' into feat/omp_target_is_accessible

nicebert committed:
[OpenMP] Update OpenMP support document

Marks omp_target_is_accessible routine as done.

nicebert committed:
Merge branch 'feat/omp_target_is_accessible' of github.com:nicebert/llvm-project into feat/omp_target_is_accessible
